draft-ietf-tls-tls13.md

```diff
@@ -2950,6 +2956,120 @@ complete picture is that ECDH is employed with a non-trivial KDF
 because TLS does not directly use the premaster secret for anything
 other than for computing the master secret.)
 
+## Update Protocol
+
+TLS supports an Update protocol which allows endpoints to update the
+cryptographic state of the connection after the handshake has been
+completed. The Update protocol can currently be used for the following tasks:
+
+- Refreshing cryptographic keying material.
+
+- Establishing a shared key which can be used to resume future
+  sessions.
+
+In versions of TLS prior to TLS 1.3, these functions were part of the handshake
+and/or required renegotiation. In TLS 1.3 they have been moved into a separate
+subprotocol. Future versions of this specification may add new functions to
+the Update protocol.
+
+
+### General Update Protocol Behavior
+
+In general, Update messages have the following structure:
+
+       enum { rekey(1), ack(2),  authentication(3), (255) } UpdateType;
+
+       struct {
+           UpdateType msg_type;       /* update type */
+           uint24 length;             /* bytes in message */
+
+           switch (msg_type) {
+             case ack:
+               UpdateAck ack;
+ 
+             case rekey:
+               UpdateRekey rekey;
+  
+             /* This structure may be extended */
+           }
+        } Update;
+
+Whenever a site sends an Update message, it updates its own master secret
+as follows:
+
+      X_master_secret_{i+1} = PRF(X_master_secret_i, "master_secret_update",
+                                  Update_message)
+
+This new master secret is used to generate traffic keys for subsequent
+records that it sends. Similarly, when a site receives an Update
+message, it updates its view of the peer's master secret and uses that
+for new traffic that it receives from the peer. The previous
+master secret MUST be discarded. [[TODO: DTLS?]]
```

Comment:
I think that we need to talk about traffic keys and master secret separately here.  That is, you hold the master secret that will be used to generate the *next* master secret and traffic keys, plus the current set of traffic keys.  The current traffic keys are derived from the *last* master secret (which has since been discarded).